backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts

```diff
@@ -1,21 +1,135 @@
-import { IsNotEmpty, IsDefined } from 'class-validator';
-import { MARKED_DECISION_POINT_STATUS, PAYLOAD_TYPE } from 'upgrade_types';
+import { Type } from 'class-transformer';
+import { IsNotEmpty, IsDefined, IsString, IsOptional, IsEnum, IsObject, ValidateIf, ValidateNested, ValidationOptions, registerDecorator } from 'class-validator';
+import { EXPERIMENT_TYPE, MARKED_DECISION_POINT_STATUS, PAYLOAD_TYPE } from 'upgrade_types';
+
+export const IsAssignedFactorRecord = (validationOptions?: ValidationOptions) => {
+  return function (object: unknown, propertyName: string) {
+    registerDecorator({
+      name: 'IsAssignedFactorRecord',
+      target: object.constructor,
+      propertyName: propertyName,
+      constraints: [],
+      options: {
+        message: 'The assignedFactor is not a valid Record<string, AssignedFactor>',
+        ...validationOptions,
+      },
+      validator: {
+        validate(value: any) {
+          return validateAssignedFactorData(value)
+        },
+      },
+    });
+  };
+};
+
+function validateAssignedFactorData(
+  data: any
+): boolean {
+  const keys = Object.keys(data);
+  for (const key of keys) {
+    const factor = data[key];
+    if (!isValidAssignedFactor(factor)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function isValidAssignedFactor(value: any): value is AssignedFactor {
+  return (
+    typeof value === 'object' &&
+    typeof value.level === 'string' &&
+    (value.payload === null || isValidPayload(value.payload))
+  );
+}
+
+function isValidPayload(value: any): value is Payload {
+  return (
+    typeof value === 'object' &&
+    Object.values(PAYLOAD_TYPE).includes(value.type) &&
+    typeof value.value === 'string'
+  );
+}
+class Payload {
+  @IsEnum(PAYLOAD_TYPE)
+  type: PAYLOAD_TYPE;
+
+  @IsString()
+  value: string;
+}
+
+class AssignedFactor {
+  @IsDefined()
+  @IsString()
+  level: string;
+
+  @ValidateIf((object, value) => value !== null)
+  @IsObject()
+  @ValidateNested()
+  @Type(() => Payload)
+  payload: Payload | null;
```

Comment:
Can you use isOptional here?

Reply:
Actually, we don't give payload as an optional value in the code. We pass the payload object or null if not found.